Sources/App/Commands/ReconcilePackageListCommand.swift

```diff
@@ -0,0 +1,116 @@
+import Vapor
+import Fluent
+import FluentPostgreSQL
+
+final class ReconcilePackageListCommand: Command
+{
+  static let name = "reconcile_package_list"
+  
+  var arguments: [CommandArgument] = []
+  var options: [CommandOption] = []
+  var help = ["Synchronise the database with the master package list from GitHub."]
+  
+  func run(using context: CommandContext) throws -> Future<Void>
+  {
+    // Grab the GitHub master package list and the database's current package list
+    let masterPackageList = try fetchMasterPackageList(context)
+    let currentPackageList = try fetchCurrentPackageList(context)
+    
+    // Reconcile the package lists and add/remove any new/deleted packages
+    return masterPackageList.and(currentPackageList).flatMap { (master, current) in
+      let reconciler = PackageListReconciler(masterPackageList: master, currentPackageList: current)
+      
+      return context.container.withPooledConnection(to: .psql) { database in
+        // Add any new packages
+        let packageAdditions = reconciler.packagesToAdd.map { url -> Future<Package> in
+          let package = Package(url: url)
+          return package.create(on: database)
+        }
+        
+        // Delete any removed packages
+        let packageDeletions = reconciler.packagesToDelete.map { url in
+          return Package.findByUrl(on: database, url: url).map { package -> Future<Void> in
+            print("Deleting \(package.url)")
+            return package.delete(on: database).map {
+              print("Deleted \(package.url)")
+            }
+          }
+        }
+
+        // Queue up all of the additions and deletions for execution
+        return packageAdditions.flatten(on: database)
+          .and(packageDeletions.flatten(on: database))
+          .transform(to: ())
+      }
+    }
+  }
+  
+  func fetchMasterPackageList(_ context: CommandContext) throws -> Future<[URL]>
+  {
+    return try context.container.client()
+      .get(masterPackageListURL)
+      .flatMap(to: [String].self) { response in
+        // GitHub serves the file as text/plain so it needs to be forced as JSON
+        response.http.contentType = .json
+        return try response.content.decode([String].self)
+    }.flatMap { urlStrings in
+      var urls = [URL]()
+      var errorNotifications = [Future<Void>]()
+
+      for urlString in urlStrings {
+        // This should delete/re-insert a few (5 or so) random packages per run of this command.
+        // ---
+        // This is just debugging code I put in to exercise the creation/deletion code. I was going
+        // to suggest that `ReconcilePackageListCommand` didn't really need testing since the logic
+        // gets tested in `PackageListReconciler`, but the fact that nothing is being deleted makes
+        // that look a little foolish.
+        if Int.random(in: 1...700) == 1 {
+          continue
+        }
```

Comment:
So I was going to suggest that this class didn't really need tests as the main logic is encapsulated within the `PackageListReconciler` class (which is tested) but then these deletes just don't execute, so I think this might need tests.

I haven't yet implemented them as I'm not sure how to mock the network response of the master package list. Any help here is appreciated.

Reply:
In tests to mock network responses you can create a `FakeClient` and then choose to use that with `config.prefer(FakeClient.self, for: Client.self)` - I have some examples I can dig up if you want